client/transaction-pool/src/metrics.rs

```diff
@@ -0,0 +1,47 @@
+// Copyright 2020 Parity Technologies (UK) Ltd.
+// This file is part of Substrate.
+
+// Substrate is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// Substrate is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with Substrate.  If not, see <http://www.gnu.org/licenses/>.
+
+//! Transaction pool prometheus metrics.
+
+use prometheus_endpoint::{register, Counter, Gauge, PrometheusError, Registry, U64};
+
+/// Transaction pool prometheus metrics.
+pub struct Metrics {
+	pub validation_pending: Gauge<U64>,
+	pub total_validated: Counter<U64>,
```

Comment:
I think it would be nice to have some median over the time it takes to validate a transaction?

Reply:
Probably, I will think of what I need most while resolving current issues or others to come

Reply:
I have a bit of experience with Prometheus now.

In my opinion, `validation_pending` should be replaced with a `validations_started` of type `Counter` and counts the number of validations that have started. Then you can do `total_validated - validation_started` to see the number of pending validations.

Also, I don't know how difficult it is to implement that, but `total_validated` could instead be a `Histogram`, and you would call `observe()` with the time it took to validate.
Histograms don't just report the time it took but also the number of times `observe()` has been called, so this can serve as a counter for the number of finished validations.

Reply:
Instead of exposing a `median` I would suggest exposing a Prometheus Histogram. Medians might be misleading as they can hide long tail latencies.

Reply:
hmm.. Thanks for heads up about Gauges, probably saved me some time. 
I thought even metrics are scrapped every couple of seconds, they also scrap all changes that were made between scraps, not just immediate value, but according to docs this seems to not be the case.

I will change to counter.

As for time, I'm not interested in it currently, but will probably add later.

Reply:
> they also scrap all changes that were made between scraps

No, this is a bit of a pitfall at the beginning. Prometheus is making a tradeoff between resolution and performance favoring performance in this place. E.g. recording every single change would require an allocated vector whereas only taking snapshots enables one to use *cheap* atomic variables.